File: csharp/src/Ice/Internal/IdleTimeoutTransceiverDecorator.cs
// Copyright (c) ZeroC, Inc.

#nullable enable

using System.Diagnostics;
using System.Net.Sockets;

namespace Ice.Internal;

#pragma warning disable CA1001 // _readTimer and _writeTimer are disposed by destroy.
internal sealed class IdleTimeoutTransceiverDecorator : Transceiver
#pragma warning restore CA1001
{
    private readonly Transceiver _decoratee;
    private readonly TimeSpan _idleTimeout;
    private readonly System.Threading.Timer? _readTimer;
    private readonly System.Threading.Timer _writeTimer;

    // Called by ConnectionI with its mutex locked.
    internal bool idleCheckEnabled { get; private set; }

    public override string ToString() => _decoratee.ToString()!;

    public EndpointI? bind() => _decoratee.bind();

    public void checkSendSize(Buffer buf) => _decoratee.checkSendSize(buf);

    public void close()
    {
        disableIdleCheck();
        cancelWriteTimer();
        _decoratee.close();
    }

    public int closing(bool initiator, LocalException ex) => _decoratee.closing(initiator, ex);

    public void destroy()
    {
        disableIdleCheck();
        cancelWriteTimer();
        _decoratee.destroy();
        _readTimer?.Dispose();
        _writeTimer.Dispose();
    }

    public Socket fd() => _decoratee.fd();

    public void finishRead(Buffer buf)
    {
        if (idleCheckEnabled)
        {
            // We don't want the idle check to run while we're reading, so we reschedule it before reading.
            rescheduleReadTimer();
        }
        _decoratee.finishRead(buf);
    }

    // We call write after finishWrite, so no need to do anything here.
    public void finishWrite(Buffer buf) => _decoratee.finishWrite(buf);

    public ConnectionInfo getInfo(bool incoming, string connectionId, string adapterName) =>
        _decoratee.getInfo(incoming, connectionId, adapterName);

    public int initialize(Buffer readBuffer, Buffer writeBuffer, ref bool hasMoreData) =>
        _decoratee.initialize(readBuffer, writeBuffer, ref hasMoreData);

    public string protocol() => _decoratee.protocol();

    public int read(Buffer buf, ref bool hasMoreData)
    {
        // We don't want the idle check to run while we're reading, so we reschedule it before reading.
        if (idleCheckEnabled)
        {
            rescheduleReadTimer();
        }
        return _decoratee.read(buf, ref hasMoreData);
    }

    public void setBufferSize(int rcvSize, int sndSize) => _decoratee.setBufferSize(rcvSize, sndSize);

    // We always call finishRead or read to actually read the data.
    public bool startRead(Buffer buf, AsyncCallback callback, object state) =>
        _decoratee.startRead(buf, callback, state);

    // startWrite is called right after write returns SocketOperationWrite. The logic in write is sufficient.
    public bool startWrite(Buffer buf, AsyncCallback callback, object state, out bool completed) =>
        _decoratee.startWrite(buf, callback, state, out completed);

    public string toDetailedString() => _decoratee.toDetailedString();

    public int write(Buffer buf)
    {
        cancelWriteTimer();
        int op = _decoratee.write(buf);
        if (op == SocketOperation.None) // write completed
        {
            rescheduleWriteTimer();
        }
        return op;
    }

    internal IdleTimeoutTransceiverDecorator(
        Transceiver decoratee,
        ConnectionI connection,
        TimeSpan idleTimeout,
        bool enableIdleCheck)
    {
        Debug.Assert(idleTimeout > TimeSpan.Zero);

        _decoratee = decoratee;
        _idleTimeout = idleTimeout;

        if (enableIdleCheck)
        {
            _readTimer = new System.Threading.Timer(_ => connection.idleCheck(_idleTimeout));

            // idleCheckEnabled remains false for now. ConnectionI calls enableIdleCheck() when it becomes Active to
            // schedule the first idle check and set idleCheckEnabled to true.
            // We don't want the idle check to start when a client connection is waiting for the initial
            // ValidateConnection message or when a server connection is in its initial StateHolding state.
        }
        _writeTimer = new System.Threading.Timer(_ => connection.sendHeartbeat());
    }

    internal void enableIdleCheck()
    {
        if (!idleCheckEnabled && _readTimer is not null)
        {
            rescheduleReadTimer();
            idleCheckEnabled = true;
        }
    }

    internal void disableIdleCheck()
    {
        if (idleCheckEnabled && _readTimer is not null)
        {
            cancelReadTimer();
            idleCheckEnabled = false;
        }
    }

    internal void scheduleHeartbeat() => rescheduleWriteTimer();

    private void cancelReadTimer() => _readTimer!.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);

    private void cancelWriteTimer() => _writeTimer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);

    private void rescheduleReadTimer() => _readTimer?.Change(_idleTimeout, Timeout.InfiniteTimeSpan);

    private void rescheduleWriteTimer() => _writeTimer.Change(_idleTimeout / 2, Timeout.InfiniteTimeSpan);
}

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/IdleTimeoutTransceiverDecorator.java
// Copyright (c) ZeroC, Inc.

package com.zeroc.IceInternal;

import com.zeroc.Ice.ConnectionI;
import com.zeroc.Ice.LocalException;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

// Decorates Transceiver to send heartbeats and optionally detect when no byte is received/read for
// a while.
// This decorator must not be applied on UDP connections.
public class IdleTimeoutTransceiverDecorator implements Transceiver {
  private final Transceiver _decoratee;
  private final int _idleTimeout;
  private final ScheduledExecutorService _scheduledExecutorService;

  private final Runnable _idleCheck;
  private final Runnable _sendHeartbeat;

  // All calls to a transceiver are serialized by the parent ConnectionI's mutex.
  private ScheduledFuture<?> _readTimer;
  private ScheduledFuture<?> _writeTimer;

  @Override
  public java.nio.channels.SelectableChannel fd() {
    return _decoratee.fd();
  }

  @Override
  public void setReadyCallback(ReadyCallback callback) {
    _decoratee.setReadyCallback(callback);
  }

  @Override
  public int initialize(Buffer readBuffer, Buffer writeBuffer) {
    int op = _decoratee.initialize(readBuffer, writeBuffer);

    if (op == SocketOperation.None) { // connected
      rescheduleReadTimer();
      rescheduleWriteTimer();
    }

    return op;
  }

  @Override
  public int closing(boolean initiator, LocalException ex) {
    return _decoratee.closing(initiator, ex);
  }

  @Override
  public void close() {
    cancelReadTimer();
    cancelWriteTimer();
    _decoratee.close();
  }

  @Override
  public EndpointI bind() {
    return _decoratee.bind();
  }

  @Override
  public int write(Buffer buf) {
    cancelWriteTimer();
    int op = _decoratee.write(buf);
    if (op == SocketOperation.None) { // write completed
      rescheduleWriteTimer();
    }
    return op;
  }

  @Override
  public int read(Buffer buf) {
    // We don't want the idle check to run while we're reading, so we reschedule it before reading.
    rescheduleReadTimer();
    return _decoratee.read(buf);
  }

  @Override
  public String protocol() {
    return _decoratee.protocol();
  }

  @Override
  public String toString() {
    return _decoratee.toString();
  }

  @Override
  public String toDetailedString() {
    return _decoratee.toDetailedString();
  }

  @Override
  public com.zeroc.Ice.ConnectionInfo getInfo() {
    return _decoratee.getInfo();
  }

  @Override
  public void checkSendSize(Buffer buf) {
    _decoratee.checkSendSize(buf);
  }

  @Override
  public void setBufferSize(int rcvSize, int sndSize) {
    _decoratee.setBufferSize(rcvSize, sndSize);
  }

  public IdleTimeoutTransceiverDecorator(
      Transceiver decoratee,
      ConnectionI connection,
      int idleTimeout,
      boolean enableIdleCheck,
      ScheduledExecutorService scheduledExecutorService) {
    _decoratee = decoratee;
    _idleTimeout = idleTimeout;
    _scheduledExecutorService = scheduledExecutorService;

    _idleCheck = enableIdleCheck ? () -> connection.idleCheck(idleTimeout) : null;
    _sendHeartbeat = () -> connection.sendHeartbeat();
  }

  private void cancelReadTimer() {
    if (_readTimer != null) {
      _readTimer.cancel(false);
      _readTimer = null;
    }
  }

  private void cancelWriteTimer() {
    if (_writeTimer != null) {
      _writeTimer.cancel(false);
      _writeTimer = null;
    }
  }

  private void rescheduleReadTimer() {
    if (_idleCheck != null) {
      cancelReadTimer();
      _readTimer = _scheduledExecutorService.schedule(_idleCheck, _idleTimeout, TimeUnit.SECONDS);
    }
  }

  private void rescheduleWriteTimer() {
    cancelWriteTimer();
    _writeTimer =
        _scheduledExecutorService.schedule(
            _sendHeartbeat, _idleTimeout * 1000 / 2, TimeUnit.MILLISECONDS);
  }
}
